cusynchronization.jl

```diff
@@ -0,0 +1,82 @@
+module CuSynchronization
+
+using CUDAnative.LLVM
+using CUDAnative.LLVM.Interop
+using CUDAnative
+
+@generated function unsafe_volatile_load(ptr::Ptr{T}) where T
+    eltyp = convert(LLVMType, T)
+    T_ptr = convert(LLVMType, Ptr{T})
+    T_actual_ptr = LLVM.PointerType(eltyp)
+
+    # create a function
+    param_types = [T_ptr]
+    llvm_f, _ = create_function(eltyp, param_types)
+
+    Builder(JuliaContext()) do builder
+        entry = BasicBlock(llvm_f, "entry", JuliaContext())
+        position!(builder, entry)
+
+        ptr = inttoptr!(builder, parameters(llvm_f)[1], T_actual_ptr)
+        ld = load!(builder, ptr)
+        LLVM.API.LLVMSetVolatile(LLVM.ref(ld), LLVM.True)
+        ret!(builder, ld)
+    end
+
+    call_function(llvm_f, T, Tuple{Ptr{T}}, :(ptr,))
+end
+
+@generated function unsafe_volatile_store!(ptr::Ptr{T}, val::T) where T
+    eltyp = convert(LLVMType, T)
+    T_ptr = convert(LLVMType, Ptr{T})
+    T_actual_ptr = LLVM.PointerType(eltyp)
+
+    # create a function
+    param_types = [T_ptr, eltyp]
+    llvm_f, _ = create_function(LLVM.VoidType(JuliaContext()), param_types)
+
+    Builder(JuliaContext()) do builder
+        entry = BasicBlock(llvm_f, "entry", JuliaContext())
+        position!(builder, entry)
+
+        ptr = inttoptr!(builder, parameters(llvm_f)[1], T_actual_ptr)
+        val = parameters(llvm_f)[2]
+        st = store!(builder, val, ptr)
+        LLVM.API.LLVMSetVolatile(LLVM.ref(st), LLVM.True)
+        ret!(builder)
+    end
+
+    call_function(llvm_f, Cvoid, Tuple{Ptr{T}, T}, :(ptr, val))
+end
+
+
+import CUDAnative: DevicePtr
+
+struct Semaphore{T, AS}
+    sem::DevicePtr{T, AS}
+    value::T
+end
+
+@inline function Base.getindex(sem::Semaphore{T}) where T
+    ptr = Base.unsafe_convert(Ptr{T}, sem.sem)
+    unsafe_volatile_load(ptr)
+end
+
+@inline function Base.setindex!(sem::Semaphore{T}, val) where T
+    ptr = Base.unsafe_convert(Ptr{T}, sem.sem)
+    unsafe_volatile_store!(ptr, convert(T, val))
+end
+
+@inline function wait(sem::Semaphore{T}) where T
+    while true
+        if sem[] == sem.value
+            break
+        end
+        threadfence_block()
```

Comment:
@maleadt I am not sure that these `threadfence_block` calls are necessary since no other thread on the device is listening to flag. The code I was inspired by does have them https://github.com/gpudirect/libgdsync/blob/3d0333d6d57af2794088c74f088723f509f794d3/include/gdsync/device.cuh#L112 but they are doing synchronization for another PCI-E device.